### PR TITLE
Conversion Of Bot Task Module with App Task Module Csharp Sample

### PR DESCRIPTION
**Pull request by Mohammed-MSFT**

### CODE REVIEW

File: samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        public static Attachment GetAdaptiveCard()
+        {
+            // Parse the JSON 
+            AdaptiveCardParseResult result = AdaptiveCard.FromJson(GetAdaptiveCardJson());
+
+            return new Attachment()
+            {
+                ContentType = AdaptiveCard.ContentType,
+                Content = result.Card
+
```
Comment (Pawan-MSFT) [inline samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs:47]: remove additional spaces

File: samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /// <summary>
+    ///  Helper class which posts to the saved channel every 20 seconds.
+    /// </summary>
+    public static class AdaptiveCardHelper
+    {
+        public static Attachment GetAdaptiveCard()
+        {
+            // Parse the JSON 
+            AdaptiveCardParseResult result = AdaptiveCard.FromJson(GetAdaptiveCardJson());
+
```
Comment (Pawan-MSFT) [inline samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs:42]: remove additional spaces

File: samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs (hunk 1 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs:47]: fixed

File: samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs (hunk 2 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs:42]: fixed

File: samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs (hunk 1 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs:47]: [Fixed](image)

File: samples/bot-task-module/csharp/Pages/PowerApp.cshtml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+﻿@page
+@using Microsoft.Extensions.Configuration
+@inject IConfiguration Configuration
+@{
+    ViewData["Title"] = "PowerApp";
+    Layout = "~/Pages/Shared/_Layout.cshtml";
+}
+
+<iframe width="720" height="520" style="width: 94%;" src="https://web.powerapps.com/webplayer/iframeapp?source=iframe&amp;screenColor=rgba(104,101,171,1)&amp;appId=/providers/Microsoft.PowerApps/apps/f998c01d-522c-46ad-9690-f67f271afdfe"></iframe>
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/Pages/PowerApp.cshtml:9]: Do we need this id?

File: samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs:1]: // Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

File: samples/bot-task-module/csharp/Helper/ApplicationSettings.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/Helper/ApplicationSettings.cs:1]: // Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

File: samples/bot-task-module/csharp/Helper/DeeplinkHelper.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/Helper/DeeplinkHelper.cs:1]: // Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

File: samples/bot-task-module/csharp/Helper/UIConstants.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED ""AS IS"", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/Helper/UIConstants.cs:22]: // Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

File: samples/bot-task-module/csharp/Helper/UIConstants.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Web;
+
+namespace Microsoft.Teams.Samples.TaskModule.Web.Helper
+{
+
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/Helper/UIConstants.cs:32]: Remove space

File: samples/bot-task-module/csharp/Pages/PowerApp.cshtml (hunk 3 above)
Comment (ChetanSharma-msft) [reply samples/bot-task-module/csharp/Pages/PowerApp.cshtml:9]: Add a place holder like <PowerBI-App-Id> and explain it in Readme for update and how to get this id?

File: samples/bot-task-module/csharp/Pages/PowerApp.cshtml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        This task module shows an "Asset Checkout" PowerApp (you'll find more information on it <a href="https://powerusers.microsoft.com/t5/Community-Apps-Gallery/Asset-Checkout-Microsoft-sample/td-p/43221" target="_blank">here</a>). Unfortunately, PowerApps
+        are tenant-specific. The one in this sample only works on the Microsoft.com tenant.
+    </p>
+    <p>To create your own:</p>
+    <ol>
+        <li>Go to <a href="https://powerapps.microsoft.com" target="_blank">https://powerapps.microsoft.com</a> and sign in.</li>
+        <li>There's a search box in the middle of the page: search for "Asset Checkout".</li>
+        <li>Click on it, press the "Make this app" button, and save it to the cloud.</li>
+        <li>Press the Share button and you'll see a page with the newly-created PowerApp's App ID. Copy it to the clipboard.</li>
+        <li>Replace "a3447344-9220-4868-b45b-07759637be0f" in the <code>src=</code> URL of the &lt;iframe&gt; in this file (Microsoft.Teams.Samples.TaskModule.Web/Views/Home/PowerApp.cshtml) with your App ID.</li>
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/Pages/PowerApp.cshtml:21]: Update it.

File: samples/bot-task-module/csharp/appsettings.json [hunk 10]
```diff
@@ -1,7 +1,7 @@
...
-  "MicrosoftAppPassword": "",
-  "MicrosoftAppTenantId": "",
-  "BaseUrl": "https://YourDeployedBotUrl.com"
-}
+  "MicrosoftAppType": "<<MicrosoftAppType>>",
+  "MicrosoftAppId": "<<MicrosoftAppId>>",
+  "MicrosoftAppPassword": "<<MicrosoftAppPassword>>",
+  "MicrosoftAppTenantId": "<<MicrosoftAppTenantId>>",
+  "BaseUrl": "<<Ngrok>>"
+}
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/appsettings.json:7]: Update the Readme for these updated placeholders.

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Added my review comments.

File: samples/bot-task-module/csharp/Pages/PowerApp.cshtml (hunk 3 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Pages/PowerApp.cshtml:9]: Comment Solved

File: samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs (hunk 4 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/AdaptiveCardHelper.cs:1]: fixed

File: samples/bot-task-module/csharp/Helper/ApplicationSettings.cs (hunk 5 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/ApplicationSettings.cs:1]: fixed

File: samples/bot-task-module/csharp/Helper/UIConstants.cs (hunk 7 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/UIConstants.cs:22]: fixed

File: samples/bot-task-module/csharp/Helper/UIConstants.cs (hunk 8 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/UIConstants.cs:32]: fixed

File: samples/bot-task-module/csharp/Pages/PowerApp.cshtml (hunk 9 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Pages/PowerApp.cshtml:21]: fixed

File: samples/bot-task-module/csharp/appsettings.json (hunk 10 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/appsettings.json:7]: fixed

**Review by Pawan-MSFT — CHANGES_REQUESTED** [review]
Please confirm my feedbacks done, if so then resolve it.

File: samples/bot-task-module/csharp/Helper/ApplicationSettings.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+using System.Collections.Generic;
+using System.Configuration;
+using System.Web;
+
+namespace Microsoft.Teams.Samples.TaskModule.Web.Helper
+{
+    public  class ApplicationSettings
+    {
+        public string BaseUrl { get; set; }
+
```
[the remark this answers is not in the document]
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/ApplicationSettings.cs:14]: fixed

File: samples/bot-task-module/csharp/Helper/ApplicationSettings.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        public string MicrosoftAppId { get; set; }
+
```
[the remark this answers is not in the document]
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/ApplicationSettings.cs:16]: fixed

File: samples/bot-task-module/csharp/Helper/DeeplinkHelper.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    public  class DeeplinkHelper
+    {
+        public  string DeepLink { get; set; }
+        public  string DeepLinkToAdaptiveCard { get; set; }
+
+        public DeeplinkHelper(string MicrosoftAppId ,string BaseUrl)
+        {
+            MicrosoftAppId = MicrosoftAppId.Replace('"', ' ').Trim();
+            BaseUrl = BaseUrl.Replace('"', ' ').Trim();
+
```
[the remark this answers is not in the document]
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Helper/DeeplinkHelper.cs:18]: fixed

File: samples/bot-task-module/csharp/Pages/Configure.cshtml [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        <div>
+            <input type="radio" name="tabType" value="tasks" onclick="onClick()">Task Module<br>
+            <input type="radio" name="tabType" value="helloworld" onclick="onClick()"> Hello World
+        </div>
+    </p>
+</form>
+<script type="text/javascript">
+    microsoftTeams.initialize();
+    microsoftTeams.settings.registerOnSaveHandler(function (saveEvent) {
+
```
[the remark this answers is not in the document]
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Pages/Configure.cshtml:23]: fixed

File: samples/bot-task-module/csharp/Pages/Tasks.cshtml [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+<script src="~/Scripts/teamsapp.js"></script>
+<script>
+    microsoftTeams.initialize();
+
+    submitHandler = (err, result) => {
+            console.log(`Submit handler - err: ${err}`);
+            alert("Result = " + JSON.stringify(result) + "\nError = " + JSON.stringify(err));
+        };
+
+        function submit(type) {
```
[the remark this answers is not in the document]
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Pages/Tasks.cshtml:49]: fixed

File: samples/bot-task-module/csharp/TeamsAppManifest/manifest.json [hunk 16]
```diff
@@ -15,14 +15,34 @@
     "outline": "icon-outline.png"
   },
   "name": {
-    "short": "Task Module",
+    "short": "Task Module_Conver",
```
[the remark this answers is not in the document]
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/TeamsAppManifest/manifest.json:18]: fixed Comments

File: samples/bot-task-module/csharp/Pages/Tasks.cshtml [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        <a target="_blank" href="@(deepLinkHelper.DeepLink)">Deep link to Custom Form page</a>
+    </div>
+<p></p>
+<script src="~/Scripts/teamsapp.js"></script>
+<script>
+    microsoftTeams.initialize();
+
+    submitHandler = (err, result) => {
+            console.log(`Submit handler - err: ${err}`);
+            alert("Result = " + JSON.stringify(result) + "\nError = " + JSON.stringify(err));
```
[the remark this answers is not in the document]
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/Pages/Tasks.cshtml:46]: Alert Removed Comment Fixed.

File: samples/bot-task-module/csharp/README.md [hunk 18]
```diff
@@ -61,10 +62,10 @@ the Teams service needs to call into the bot.
...
-    - Set "MicrosoftAppType" in the `appsettings.json`. (**Allowed values are: MultiTenant(default), SingleTenant, UserAssignedMSI**)
+1) Update the `appsettings.json` configuration for the bot to use the `<<MicrosoftAppId>>`, `<<MicrosoftAppPassword>>`, `<<MicrosoftAppTenantId>>` generated in Step 2 (App Registration creation). (Note the App Password is referred to as the "client secret" in the azure portal and you can always create a new client secret anytime.)
+    - Set `<<MicrosoftAppType>>` in the `appsettings.json`. (**Allowed values are: MultiTenant(default), SingleTenant, UserAssignedMSI**)
 
-    - Set "BaseUrl" in the `appsettings.json` as per your application like the ngrok forwarding url (ie `https://xxxx.ngrok.io`) after starting ngrok
+    - Set `<<BaseUrl>>` in the `appsettings.json` as per your application like the ngrok forwarding url (ie `https://xxxx.ngrok.io`) after starting ngrok
```
Comment (ChetanSharma-msft) [inline samples/bot-task-module/csharp/README.md:68]: You did not mention valid-domain update steps in readme for manifest but place holder is also differnt:
image

**Review by ChetanSharma-msft — APPROVED** [review]
Approved with suggestions.

File: samples/bot-task-module/csharp/README.md (hunk 18 above)
Comment (Mohammed-MSFT) [reply samples/bot-task-module/csharp/README.md:68]: Updated Domain Name in Manifest Section

**Review by Pawan-MSFT — APPROVED** [review]
looks correct, approving!

**Review by Wajeed-msft — APPROVED** [review]

**Comment by ChetanSharma-msft** [comment]
Executed build pipeline: https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/3805373664

**Pull request merged**